### PR TITLE
Fix XSS on activites and throghout the rest of app [SCI-102]

### PR DESCRIPTION
**Pull request by okriuchykhin**
https://biosistemika.atlassian.net/browse/SCI-102

Sanitize is done on input and output through helper methods and whitelists of tags and attributes.
We need pay attention and do not sanitize html code with data attributes as they are not whitelisted.

### CODE REVIEW

**Comment by mz3944** [comment]
A side note: if you've opened also a PR in some addon/s, then mention it in teh PR description (because I almost forgot you've also made changes to billing and gamification modules).

**Comment by okriuchykhin** [comment]
> app/helpers/protocol_status_helper.rb:
> def protocol_status_popover_title(protocol) - add sanitize_input(res) at the end of this method, like you did everywhere else, and sanitize_input(protocol.added_by.full_name) is redundant then

We cannot sanitize whole block as it contains data attributes, so we just sanitize variables

**Comment by okriuchykhin** [comment]
> app/controllers/reports_controller.rb
> 491- res - replace this line with sanitize_input(res)
> 508- res - replace this line with sanitize_input(res)
> 572- res - replace this line with sanitize_input(res)
> 612- res - replace this line with sanitize_input(res)
> 626- res - replace this line with sanitize_input(res)

res is array there, so can't call sanitize on it. But all elements goes through render_to_string, so everything should be already escaped

**Comment by mz3944** [comment]
Replace every _auto_link_ call with custom _auto_link_input_ function, defined in _InputSanitizeHelper_ module. It should call _auto_link_ internally, but with the whitelisted tags and attributes, e.g.:

def auto_link_input:
  auto_link(sanitize_input(step.description),
                 link: :urls,
                 html: { target: '_blank' },
                 sanitize: false).html_safe
end

**Comment by okriuchykhin** [comment]
Added auto_link wrapper with custom sanitization

File: app/views/experiments/canvas.html.erb [hunk 1]
```diff
@@ -1,4 +1,4 @@
-<% provide(:head_title, raw(t("experiments.canvas.head_title", project: @project.name))) %>
+<% provide(:head_title, sanitize_input(t("experiments.canvas.head_title", project: @project.name))) %>
```
Comment (mz3944) [inline app/views/experiments/canvas.html.erb]: If we're being pedantic, than @project.name should be escape_input(@project.name), e.g. `sanitize_input(t("experiments.canvas.head_title", project: escape_input(@project.name)))`

File: app/views/experiments/module_archive.html.erb [hunk 2]
```diff
@@ -1,4 +1,4 @@
-<% provide(:head_title, raw(t("experiments.module_archive.head_title", experiment: @experiment.name))) %>
+<% provide(:head_title, sanitize_input(t("experiments.module_archive.head_title", experiment: @experiment.name))) %>
```
Comment (mz3944) [inline app/views/experiments/module_archive.html.erb]: Same here: escape_input(@experiment.name)

File: app/views/experiments/samples.html.erb [hunk 3]
```diff
@@ -1,4 +1,4 @@
-<% provide(:head_title, raw(t("projects.samples.head_title", project: @experiment.name))) %>
+<% provide(:head_title, sanitize_input(t("projects.samples.head_title", project: @experiment.name))) %>
```
Comment (mz3944) [inline app/views/experiments/samples.html.erb]: Same here: escape_input(@experiment.name)

File: app/views/my_modules/_result.html.erb [hunk 4]
```diff
@@ -30,7 +30,7 @@
       <a class="result-panel-collapse-link" href="#result-panel-<%= result.id %>" data-toggle="collapse" data-remote="true">
         <span class="glyphicon glyphicon-collapse-down collapse-result-icon"></span>
         <strong><%= result.name %></strong> |
-        <span><%= raw t'my_modules.results.published_on', timestamp: l(result.created_at, format: :full), user: result.user.full_name %></span>
+        <span><%= sanitize_input t'my_modules.results.published_on', timestamp: l(result.created_at, format: :full), user: result.user.full_name %></span>
```
Comment (mz3944) [inline app/views/my_modules/_result.html.erb]: Same here: escape_input(result.user.full_name)

File: app/views/my_modules/activities.html.erb [hunk 5]
```diff
@@ -1,4 +1,4 @@
-<% provide(:head_title, raw(t("my_modules.activities.head_title", project: @my_module.experiment.project.name, module: @my_module.name))) %>
+<% provide(:head_title, sanitize_input(t("my_modules.activities.head_title", project: @my_module.experiment.project.name, module: @my_module.name))) %>
```
Comment (mz3944) [inline app/views/my_modules/activities.html.erb]: Same here: escape_input(@my_module.experiment.project.name) and escape_input(@my_module.name)

File: app/views/my_modules/archive.html.erb [hunk 6]
```diff
@@ -1,4 +1,4 @@
-<% provide(:head_title, raw(t("my_modules.module_archive.head_title", project: @my_module.experiment.project.name, module: @my_module.name))) %>
+<% provide(:head_title, sanitize_input(t("my_modules.module_archive.head_title", project: @my_module.experiment.project.name, module: @my_module.name))) %>
```
Comment (mz3944) [inline app/views/my_modules/archive.html.erb]: Same here: escape_input(@my_module.experiment.project.name) and escape_input(@my_module.name)

File: app/views/my_modules/protocols.html.erb [hunk 7]
```diff
@@ -1,4 +1,4 @@
-<% provide(:head_title, raw(t("my_modules.protocols.head_title", project: @project.name, module: @my_module.name))) %>
+<% provide(:head_title, sanitize_input(t("my_modules.protocols.head_title", project: @project.name, module: @my_module.name))) %>
```
Comment (mz3944) [inline app/views/my_modules/protocols.html.erb]: And here.

File: app/views/my_modules/results.html.erb [hunk 8]
```diff
@@ -1,4 +1,4 @@
-<% provide(:head_title, raw(t("my_modules.results.head_title", project: @project.name, module: @my_module.name))) %>
+<% provide(:head_title, sanitize_input(t("my_modules.results.head_title", project: @project.name, module: @my_module.name))) %>
```
Comment (mz3944) [inline app/views/my_modules/results.html.erb]: And here.

File: app/views/my_modules/samples.html.erb [hunk 9]
```diff
@@ -1,8 +1,7 @@
-<% provide(:head_title, raw(t("my_modules.samples.head_title", project: @project.name, module: @my_module.name))) %>
+<% provide(:head_title, sanitize_input(t("my_modules.samples.head_title", project: @project.name, module: @my_module.name))) %>
```
Comment (mz3944) [inline app/views/my_modules/samples.html.erb]: And here.

File: app/views/projects/experiment_archive.html.erb [hunk 10]
```diff
@@ -1,4 +1,4 @@
-<% provide(:head_title, raw(t("projects.experiment_archive.head_title", project: @project.name))) %>
+<% provide(:head_title, sanitize_input(t("projects.experiment_archive.head_title", project: @project.name))) %>
```
Comment (mz3944) [inline app/views/projects/experiment_archive.html.erb]: And here.

File: app/views/projects/samples.html.erb [hunk 11]
```diff
@@ -1,8 +1,7 @@
-<% provide(:head_title, raw(t("projects.samples.head_title", project: @project.name))) %>
+<% provide(:head_title, sanitize_input(t("projects.samples.head_title", project: @project.name))) %>
```
Comment (mz3944) [inline app/views/projects/samples.html.erb]: And here.

File: app/views/projects/show.html.erb [hunk 12]
```diff
@@ -1,4 +1,4 @@
-<% provide(:head_title, raw(t("projects.show.head_title", project: @project.name))) %>
+<% provide(:head_title, sanitize_input(t("projects.show.head_title", project: @project.name))) %>
```
Comment (mz3944) [inline app/views/projects/show.html.erb]: And here.

File: app/views/protocols/header/_keywords_label.html.erb [hunk 13]
```diff
@@ -1,5 +1,5 @@
 <% if @protocol.protocol_keywords.count > 0 %>
-  <%= @protocol.protocol_keywords.collect{ |kw| "<strong>#{sanitize(kw.name)}</strong>" }.join(", ").html_safe %>
+  <%= sanitize_input(@protocol.protocol_keywords.collect{ |kw| "<strong>#{kw.name}</strong>" }.join(", ")) %>
```
Comment (mz3944) [inline app/views/protocols/header/_keywords_label.html.erb]: And here.

File: app/views/protocols/index/_protocol_preview_modal_body.html.erb [hunk 14]
```diff
@@ -75,7 +75,7 @@
         <div class="panel panel-default">
           <div class="panel-heading">
             <strong><%= step.name %></strong> |
-            <span><%= raw t("protocols.steps.published_on", timestamp: l(step.created_at, format: :full), user: step.user.full_name) %></span>
+            <span><%= sanitize_input t("protocols.steps.published_on", timestamp: l(step.created_at, format: :full), user: step.user.full_name) %></span>
```
Comment (mz3944) [inline app/views/protocols/index/_protocol_preview_modal_body.html.erb]: And here.

File: app/views/protocols/index/_results_modal_body.html.erb [hunk 15]
```diff
@@ -23,7 +23,7 @@
         <li>
           <span class="label label-success"><%= t("protocols.index.#{en_action}.row_success") %></span>
           <% if en_action == "restore_results" %>
-            <%= t("protocols.index.restore_results.row_#{r[:type]}", protocol: r[:name]).html_safe %>
+            <%= sanitize_input(t("protocols.index.restore_results.row_#{r[:type]}", protocol: r[:name])) %>
```
Comment (mz3944) [inline app/views/protocols/index/_results_modal_body.html.erb]: And here.

File: app/views/protocols/index/_results_modal_body.html.erb [hunk 16]
```diff
@@ -36,9 +36,9 @@
           <span class="label label-warning"><%= t("protocols.index.#{en_action}.row_renamed") %></span>
           <% protocol_str = t("protocols.index.row_renamed_html", old_name: r[:name], new_name: r[:new_name]) %>
```
Comment (mz3944) [inline app/views/protocols/index/_results_modal_body.html.erb]: And here for r[:name] and r[:new_name].

File: app/views/reports/index.html.erb [hunk 17]
```diff
@@ -1,4 +1,4 @@
-<% provide(:head_title, raw(t("projects.reports.index.head_title", project: @project.name))) %>
+<% provide(:head_title, sanitize_input(t("projects.reports.index.head_title", project: @project.name))) %>
```
Comment (mz3944) [inline app/views/reports/index.html.erb]: And here.

File: app/views/reports/new.html.erb [hunk 18]
```diff
@@ -1,4 +1,4 @@
-<% provide(:head_title, raw(t("projects.reports.new.head_title", project: @project.name))) %>
+<% provide(:head_title, sanitize_input(t("projects.reports.new.head_title", project: @project.name))) %>
```
Comment (mz3944) [inline app/views/reports/new.html.erb]: And here.

File: app/views/steps/_step.html.erb [hunk 19]
```diff
@@ -26,7 +26,7 @@
       <a class="step-panel-collapse-link" href="#step-panel-<%= step.id %>" data-toggle="collapse" data-remote="true">
         <span class="glyphicon glyphicon-collapse-down collapse-step-icon"></span>
         <strong><%= step.name %></strong> |
-        <span><%= raw t("protocols.steps.published_on", timestamp: l(step.created_at, format: :full), user: step.user.full_name) %></span>
+        <span><%= sanitize_input t("protocols.steps.published_on", timestamp: l(step.created_at, format: :full), user: step.user.full_name) %></span>
```
Comment (mz3944) [inline app/views/steps/_step.html.erb]: And here.

File: app/models/checklist.rb [hunk 20]
```diff
@@ -54,4 +56,9 @@ def self.search(user, include_archived, query = nil, page = 1)
     end
   end
 
+  private
+
+  def sanitize_fields
+    self.name = sanitize_input(name)
```
Comment (mz3944) [inline app/models/checklist.rb]: This is an exception, and should be escape_input(name), as Matjaž pointed out.

**Comment by okriuchykhin** [comment]
Now everything should be good)

**Comment by mz3944** [comment]
Semm OK to me. Merge away!! :open_hands:

**Pull request merged**